[Blackjack - step1] Submit saragrosser's mission

saragrosser opened the pull request:
This PR completes the first step of the Kotlin Blackjack project: a terminal-based simplified Blackjack game where players compete against a dealer. The game includes full dealing logic, draw rules, score calculation (with Ace logic), and win/loss/draw result handling.

**What's Included:**

-Core classes for Card, Deck, Participant, Player, and Dealer

-Game controller logic (Game.kt) to run a full game flow

-Score handling extracted into ScoreCalculator.kt

-Result logic extracted into GameLogic.kt and returned via a GameResult enum

-Console input/output using InputView and OutputView

-Unit tests 

All your feedback is welcome thank you so much!

saragrosser committed:
docs (README): create README.md file with features

saragrosser committed:
style(README) : update README.md for style purposes

saragrosser committed:
feat: added Card and Deck files

saragrosser committed:
delete (Deck): delete deck file and DeckTest to try a different approach

saragrosser committed:
feat(Participant Model): add Participant abstract class for shared player/dealer logic

saragrosser committed:
feat(Player): add player logic

-update CardTest.kt to refactor display test

saragrosser committed:
feat(dealer): add Dealer class to give card

test(dealerTest): add first test to DealerTest for card dealing

saragrosser committed:
feat(dealer): add Dealer fun askPlayerDraw

saragrosser committed:
feat(view): add InputView to handle player input

saragrosser committed:
refactor(dealer): extract input logic into InputView

saragrosser committed:
test(dealer): add test for card dealing

saragrosser committed:
feat(view): add OutputView to display player and dealer hands

saragrosser committed:
feat(game): create game file

 -initialize game and collect player names

saragrosser committed:
feat(game): allow each player to draw additional cards

saragrosser committed:
feat(game): complete OutputView with dealer draw and status methods

saragrosser committed:
feat(application): create application file to run app

saragrosser committed:
feat(game): redisplay player hands after drawing

- add dealer drawing logic and display dealer cards

- display each player's final hand and score

saragrosser committed:
feat(deck): create Deck class to handle card generation and shuffling

test(deck): create tests for DECK class

saragrosser committed:
test(deck): add unit tests for card drawing and deck size

saragrosser committed:
update(game): use extracted Deck class to manage cards

-added logic forEach player playerScore, playerBusted

saragrosser committed:
update(dealer): substract scorecalculation logic for more simplicity and readability

saragrosser committed:
feat(ScoreCalculator):create file to minimize dealer responsabilities

-add calculate and isBusted functions

saragrosser committed:
update(Output): reflect changes after creating ScoreCalculator

saragrosser committed:
update(Game): reflect changes after creating ScoreCalculator

saragrosser committed:
test(score-calculator): add unit tests for score calculation and bust logic

saragrosser committed:
feat(result): add GameResult enum to represent win, lose, and draw outcomes

saragrosser committed:
refactor(Game): replace string-based results with GameResult enum

saragrosser committed:
refactor(test): delete unused imports

saragrosser committed:
Test(dealer) add test for specific dealer rule >16 or <17

saragrosser committed:
Feat(GameLogic): create Gamelogic separate object

-Substract gamelogic from Game.kt

saragrosser committed:
Test(game-logic): add unit test for GameLogic

saragrosser committed:
docs(readme): update status and checklist in README.md

Hyeon9mak commented on README.md:
```diff
@@ -1 +1,28 @@
+A simplified terminal-based Blackjack game written in Kotlin.  
+The player or dealer with a total closest to 21—without going over—wins.  
+Includes card dealing, drawing logic, and win/loss/draw result handling.
+
+## Features
+
+### Participant
+- [x] Create `Participant` abstract class with hand management
+- [x] Implement `Card` class with display and score logic
+
+### Dealer
+- [x] Deal card to any participant
+- [x] Dealer auto-draws when score ≤ 16
+- [x] Ask players if they want to draw more cards
+- [x] Check if a participant is busted
+- [x] Calculate participant score with Ace logic
+- [x] Announce game results
+
+### Game
+- [x] Accept player names
+- [x] Deal two initial cards to all participants
+- [x] Handle player turns
+- [x] Handle dealer turn
+- [x] Display final hands and scores
+- [x] Determine and print winners using `GameLogic.getGameResult()`
```
😉 👍

Hyeon9mak commented on src/main/kotlin/blackjack/Application.kt:
```diff
@@ -0,0 +1,5 @@
+
+fun main() {
+    Game().run()
+}
```
https://medium.com/@hyeon9mak/about-no-newline-at-end-of-file-warning-on-github-b42d14dccc93 😉

saragrosser replied:
Oops! thank you!! I will be more aware of the styling conventions!

Hyeon9mak commented on src/main/kotlin/blackjack/Card.kt:
```diff
@@ -0,0 +1,38 @@
+    JACK("J", 10),
+    QUEEN("Q", 10),
+    KING("K", 10),
+    ACE("A", 11)
```
When the Blackjack app convert to Web browser image game,
fellow develop should check the `InputView`, `OutputView`, `Card`, ... a lot of classes.

How about move all view logic to View classes?
Or, u can make duplicate enum class in View class file only used for output on terminal.

```kotlin
enum class SuitView(
    val suit: Suit,
    val symbol: String,
) {
    HEART(suit = Suit.HEART, symbol = "♥"),
    // ...
}
```

saragrosser replied:
Hi!! thank you for your feedback, I have moved view symbols to SuitView and RankView to have all view logic inside the view package

Hyeon9mak commented on src/main/kotlin/blackjack/Card.kt:
```diff
@@ -0,0 +1,38 @@
+
+    fun isAce(): Boolean {
+        return rank == Rank.ACE
+    }
```
> Do not use getter, Ask to Object.

Nice work! 👍 
When the `Rank` class chagned, outer classes does not be affected!

saragrosser replied:
Thank you!!

Hyeon9mak commented on src/main/kotlin/blackjack/Dealer.kt:
```diff
@@ -0,0 +1,32 @@
+package blackjack
+
+class Dealer(private val deck: Deck) : Participant("Dealer") {
```
If the name "Dealer" needs to be changed to below image file, which class file will a fellow developer open first: `OutputView`-`Dealer`?

<img width="862" height="534" alt="image" src="https://github.com/user-attachments/assets/f0acb4f5-574d-46e2-a121-ac3b36268fb5" />

saragrosser replied:
Regarding your question about which file a developer would open if we needed to change how the "Dealer" is represented (like in the image), they would now open the `OutputView.kt` file, not `Dealer.kt`.

All display logic (such as showing the dealer’s cards or describing their actions) has been moved to the view layer. The `Dealer.kt` class only contains domain logic, so it's unaffected by presentation changes. This separation follows MVC principles and keeps the domain isolated from external changes.

Thanks again for helping me see this separation more clearly! 🙂

Hyeon9mak commented on src/main/kotlin/blackjack/Dealer.kt:
```diff
@@ -0,0 +1,32 @@
+                break
+            }
+        }
+    }
```
You’ve probably learned about the MVC(Model-View-Controller) pattern in a Lotto mission, right? 😉 

I think All HeroTechCourse missions are designed under the assumption that
'the domain doesn’t change often, but the view changes frequently.'.

Your current code, the domain and view depend on each other.
When that happens, changes in one part will keep affecting the other.

How could you improve your code so that the domain isn’t affected by changes in the view? 💭

saragrosser replied:
You're absolutely right, the original `Dealer.askPlayerDraw()` method was tightly coupling the domain logic with the view.

I’ve now decoupled all view-related responsibilities by moving those interactions to `OutputView.kt`. The domain layer is now only concerned with game logic, and doesn't directly reference any input/output. This ensures the logic can evolve independently of the view, following MVC principles more closely.

To reflect this separation more clearly, I also refactored the folder structure into:
- `model` → all domain entities (Dealer, Player, Card, etc.)
- `view` → all terminal display logic (InputView, OutputView, SuitView, etc.)
- `controller` → orchestrating logic (GameLogic, GameController)

That said, I had a follow-up question:  
From what I’ve read, when not building an app with clear API endpoints (like a web controller), some recommend using a `service` layer instead of a `controller` directory to hold the logic (e.g., `GameController`). Would that be more appropriate in this case?

Thanks again for your guidance, this step really helped me understand how to maintain a clean architecture and makes it more clear to me for future project implementations.

정말 감사합니다!

Hyeon9mak commented on src/main/kotlin/blackjack/Deck.kt:
```diff
@@ -0,0 +1,21 @@
+    fun drawCard(): Card {
+        if (cards.isEmpty()) throw IllegalStateException("Deck is empty")
+        return cards.removeFirst()
+    }
```
Niiiiiiiiiice 💯 👍

Hyeon9mak commented on src/main/kotlin/blackjack/Game.kt:
```diff
@@ -0,0 +1,60 @@
+    fun run() {
+        val playerNames = InputView.askPlayerNames()
+        val players = playerNames.map { Player(it) }
+        val deck = Deck()
+        val dealer = Dealer(deck)
+
+        OutputView.displayDealing(playerNames)
```
Also possible for the `Dealer` create and hold the `Deck` inside constructor.
It's more easier maintain the code.

saragrosser replied:
Yes, it makes sense that the Dealer is the one who creates the Deck, which also makes it unnecessary to create a Deck every time separately. Thank you!

Hyeon9mak commented on src/main/kotlin/blackjack/Game.kt:
```diff
@@ -0,0 +1,60 @@
+            players.forEach { player ->
+                dealer.giveCardTo(player)
+            }
+        }
```
How about this? 😉 
I think it can make more easier to generate test code.

```kotlin
dealer.giveTwoCardsTo(players)

"dealer give two cards to all players" {
    // ...
}
```

saragrosser replied:
You are totally right! I moved this logic to the Dealer, and now the code inside Game is more readable. Thank you!

Hyeon9mak commented on src/main/kotlin/blackjack/Game.kt:
```diff
@@ -0,0 +1,60 @@
+        players.forEach { dealer.askPlayerDraw(it) }
+
+        println()
+        players.forEach { OutputView.displayPlayerHand(it) }
```
> When the Blackjack app convert to Web browser image game

is it really the `Game`'s responsibility to call `println()`? 🤔

saragrosser replied:
Hi, you are right it is not Game responsibility, it should be in view logic so I have moved  all "println" and display logic from Game to OutputView :)

Hyeon9mak commented on src/main/kotlin/blackjack/InputView.kt:
```diff
@@ -0,0 +1,20 @@
+        val input = readln()
+        return input.split(",").map { it.trim() }.filter { it.isNotEmpty() }
+    }
+    fun askPlayerWantsToDraw(playerName: String): Boolean {
+        println("Would $playerName like to draw another card? (y for yes, n for no)")
+        return when (readln().trim().lowercase()) {
+            "y" -> true
+            "n" -> false
+            else -> {
+                println("Invalid input. Please enter 'y' or 'n'.")
```
Niiiice. `y`, `n` strings are not important things of Blackjack game.
Those are just user's input buttons. 💯 👍

saragrosser replied:
I totally agree—those input values are just a UI detail and not core to the game logic. Appreciate the confirmation! 😊

Hyeon9mak commented on src/main/kotlin/blackjack/ScoreCalculator.kt:
```diff
@@ -0,0 +1,21 @@
+package blackjack
+
+object ScoreCalculator {
+
+    fun calculate(participant: Participant): Int {
+        val hand = participant.getHand()
+        var total = hand.sumOf { it.getScore() }
+        var aces = hand.count { it.isAce() }
+
+        while (total > 21 && aces > 0) {
+            total -= 10
+            aces--
+        }
+
+        return total
```
How about ask "total score" and "it had Ace card" to `Hand`?

> Do not use getter.

saragrosser replied:
Yes, this makes total sense, and after extracting the logic to a Hand class, I needed to refactor the code even more which resulted in a more readable code! Thank you!

Hyeon9mak commented on src/test/kotlin/blackjack/DeckTest.kt:
```diff
@@ -0,0 +1,34 @@
+        deck = Deck()
+    }
+
+    @Test
+    fun `should start with 52 cards`() {
+        assertEquals(52, deck.remainingCards())
+    }
+
+    @Test
+    fun `should draw one card and decrease deck size`() {
+        deck.drawCard()
+        assertEquals(51, deck.remainingCards())
+    }
+
+    @Test
+    fun `should throw error when drawing from empty deck`() {
+        repeat(52) { deck.drawCard() }
+        assertThrows(IllegalStateException::class.java) {
+            deck.drawCard()
+        }
+    }
```
Very detail test cases. nice work! 💯

saragrosser replied:
Thank you!!

Hyeon9mak reviewed (changes requested):
Hello Sari, 
Did a great job to complete Blackjack mission step1 👍 

I focused most of my comments on the MVC pattern.
It probably won’t be too difficult for you to work through them. 😄 

Feel free to ask any questions u have!

saragrosser committed:
style(Application): add missing newline at end of file

saragrosser committed:
refactor(card): move view symbols to SuitView and RankView

- update display() to use SuitView and RankView
- create view directory to store all view related files

saragrosser committed:
refactor(game)(outputview): remove unused code

saragrosser committed:
refactor(view): move println and player hand display logic from Game to OutputView

saragrosser committed:
refactor(dealer): move Deck creation inside Dealer for better encapsulation

saragrosser committed:
refactor(dealer): add giveTwoCardsTo() to make code more readable.

saragrosser committed:
refactor(domain): extract Hand class and move score logic from Participant

saragrosser committed:
refactor(structure): refactor structure of architecture to create a MVC oriented design

saragrosser committed:
refactor(suit): erase unnecessary comments

saragrosser committed:
refactor(): erase unnecessary comments

Hyeon9mak commented on src/main/kotlin/blackjack/view/SuitView.kt:
```diff
@@ -0,0 +1,16 @@
+
+import blackjack.model.Suit
+
+enum class SuitView(val suit: Suit, val symbol: String) {
+    HEART(Suit.HEART, "♥"),
+    DIAMOND(Suit.DIAMOND, "♦"),
+    CLUB(Suit.CLUB, "♣"),
```
I think u can change access modifier of `Suit` to private.
Those only use in companion object area.

saragrosser replied:
I have refactor to make suit private and I also did rank private! Thank you!

Hyeon9mak commented on src/main/kotlin/blackjack/model/GameResult.kt:
```diff
@@ -0,0 +1,7 @@
+package blackjack.model
+
+enum class GameResult(val label: String) {
+    WIN("Win"),
+    LOSE("Lose"),
+    DRAW("Draw")
+}
```
And these labels are looks like Rank and Suit 😄

saragrosser replied:
Yes I understand now better, I have created a Gameresultview file now to separate the view from the logic! thank you!!

Hyeon9mak commented on src/main/kotlin/blackjack/model/Dealer.kt:
```diff
@@ -0,0 +1,46 @@
+package blackjack.model
+
+import blackjack.view.InputView
+import blackjack.view.OutputView
+
+class Dealer : Participant("Dealer") {
+
+    private val deck = Deck()
+
+    fun giveTwoCardsTo(participants: List<Participant>) {
+        repeat(2) {
+            giveCardTo(this)
+            participants.forEach { giveCardTo(it) }
```
model class still know view package classes. 😱 

How about reversing the direction
— instead of the model using the view, let the view use the model instead?
The controller can help coordinate that interaction in between!

Hyeon9mak replied:
And Let's think about Dealer's name.
In blackjack game, The Dealer's name is really important thing? 🤔 
Maybe, it just need only for the view layer!

saragrosser replied:
I've now removed all references to InputView and OutputView from Dealer and moved the related logic into TurnManager. The model no longer depends on any view layer, and the controller fully coordinates interactions between model and view.

Also passed the "Dealer" name explicitly into Participant to keep the display logic separate from the game logic 😄

Thank you again for the guidance!

Hyeon9mak commented on src/main/kotlin/blackjack/controller/GameLogic.kt:
```diff
@@ -0,0 +1,21 @@
+package blackjack.controller
+
+import blackjack.model.GameResult
+
+object GameLogic {
+
+    fun getGameResult(
```
> when not building an app with clear API endpoints (like a web controller), some recommend using a service layer instead of a controller directory to hold the logic (e.g., GameController). Would that be more appropriate in this case?

In the end, `Controller` and `Service` are just terms that express the idea of orchestration.
What really matters is having a solid bridge between the `Model`(`Domain`), which contains the core Blackjack rules, and the View, which handles input and output.
So honestly, any naming could work.

However, I’d like to offer one suggestion.
Many developers tend to use names like
`Controller`, `Service`, `Manager`, `Utils`, `Factory` out of habit.
As you can guess, these names often end up representing large, catch-all classes that can do anything and hold anything.
This makes it hard to guess what logic is inside without checking, causing inconvenience and increasing the risk of merge conflicts, which hurts maintainability.

Therefore, I encourage practicing with more specific and intuitive names instead of generic ones like `Controller` or `Service`.
For example… how about something like `BlackjackGame`?

Hyeon9mak commented on src/main/kotlin/blackjack/model/Hand.kt:
```diff
@@ -0,0 +1,29 @@
+
+    fun getNumberOfCards() = cards.size
+
+    fun getScore(): Int {
+        var total = cards.sumOf { it.getScore() }
+        var aces = cards.count { it.isAce() }
+
+        while (total > 21 && aces > 0) {
+            total -= 10
+            aces--
+        }
+
+        return total
+    }
+
+    fun isBusted(): Boolean = getScore() > 21
```
😉 👍

Hyeon9mak commented on src/test/kotlin/blackjack/model/HandTest.kt:
```diff
@@ -0,0 +1,37 @@
+        hand = Hand()
+    }
+
+    @Test
+    fun `should calculate total correctly without aces`() {
+        hand.add(Card(Suit.HEART, Rank.EIGHT))
+        hand.add(Card(Suit.CLUB, Rank.NINE))
+        assertEquals(17, hand.getScore())
+    }
+
+    @Test
+    fun `should count ace as 11 when total stays under 21`() {
+        hand.add(Card(Suit.SPADE, Rank.ACE))
+        hand.add(Card(Suit.DIAMOND, Rank.SEVEN))
+        assertEquals(18, hand.getScore())
+    }
+
+    @Test
+    fun `should detect hand as busted when score over 21`() {
+        hand.add(Card(Suit.HEART, Rank.KING))
+        hand.add(Card(Suit.CLUB, Rank.QUEEN))
+        hand.add(Card(Suit.SPADE, Rank.TWO))
+        assertTrue(hand.isBusted())
+    }
```
👍 And I think much better append the over 21 Ace test 😉

Hyeon9mak reviewed (approved):
Hi Sari!
I'm glad to took the feedback comments positively.

I’ve left a few additional comments and also answered your question.
You seem to have a solid grasp of everything — just a few minor tweaks needed. 😄

I’ll go ahead and move you to the next step. See you in step2!
If you have any more questions, feel free to leave a comment 😉
